# Linear Models

In this lesson, we will look at regression and classification using linear models. For regression, we will be predicting housing prices using the Boston Housing dataset. For classification, we will be predicting survival using the Titanic dataset.

We will make use of both scikit-learn and TensorFlow.

# Regression

In [1]:
# load libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

%matplotlib inline

In [2]:
# read in the Boston housing data
boston_df = pd.read_csv('assets/Boston/train.csv', index_col='ID')

In [3]:
# what columns are contained in this dataset? what are the types? are there any null values?
boston_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 333 entries, 1 to 506
Data columns (total 14 columns):
crim       333 non-null float64
zn         333 non-null float64
indus      333 non-null float64
chas       333 non-null int64
nox        333 non-null float64
rm         333 non-null float64
age        333 non-null float64
dis        333 non-null float64
rad        333 non-null int64
tax        333 non-null int64
ptratio    333 non-null float64
black      333 non-null float64
lstat      333 non-null float64
medv       333 non-null float64
dtypes: float64(11), int64(3)
memory usage: 39.0 KB


the data has 333 records, all columns are numeric, and we have no null entries. It's our lucky day

In [4]:
# take a peak at the records to see what they look like
boston_df.head()

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
ID                                                                              
1   0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
2   0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
4   0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
5   0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   
7   0.08829  12.5   7.87     0  0.524  6.012  66.6  5.5605    5  311     15.2   

     black  lstat  medv  
ID                       
1   396.90   4.98  24.0  
2   396.90   9.14  21.6  
4   394.63   2.94  33.4  
5   396.90   5.33  36.2  
7   395.60  12.43  22.9

In [5]:
boston_df.tail()

crim   zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
ID                                                                              
500  0.17783  0.0   9.69     0  0.585  5.569  73.5  2.3999    6  391     19.2   
502  0.06263  0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273     21.0   
503  0.04527  0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273     21.0   
504  0.06076  0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273     21.0   
506  0.04741  0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273     21.0   

      black  lstat  medv  
ID                        
500  395.77  15.10  17.5  
502  391.99   9.67  22.4  
503  396.90   9.08  20.6  
504  396.90   5.64  23.9  
506  396.90   7.88  11.9

# scikit-learn
our target/label is called `medv`. To use scikit-learn, we will simply split our data into two (a training set and a validation set). The normal ratio is 70% for training and 30% for validation. scikit-learn let's us create this using `train-test-split()`

Before splitting, we need to separate our `boston_df` into features and labels!

In [6]:
X = boston_df.drop('medv', axis=1) # this returns a DataFrame
y = boston_df['medv'] # this returns a Series

In [7]:
from sklearn.model_selection import train_test_split

X_train, X_val, y_train, y_val = train_test_split(X, y, random_state=42, train_size=0.3)

/anaconda3/envs/tensorflow/lib/python3.6/site-packages/sklearn/model_selection/_split.py:2026: FutureWarning: From version 0.21, test_size will always complement train_size unless both are specified.
  FutureWarning)


We now have our training and validation datasets, and are ready to train our first linear model.

We will use scikit-learn to build a Linear Regression model using all of our features.

In [8]:
from sklearn.linear_model import LinearRegression

linearModel = LinearRegression()
linearModel.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

_That was simple, we have trained a Machine Learning model!_

Is the model any good? We can find out by scoring the model.

In [9]:
linearModelScore = linearModel.score(X_val, y_val)
linearModelScore

0.6649640156673967

We can make predictions by calling `predict()` on our model

In [10]:
linearModelPredictions = linearModel.predict(X_val)

In [11]:
print(type(linearModelPredictions))

<class 'numpy.ndarray'>


In [12]:
linearModelPredictions

array([ 2.26235712e+01,  2.34212241e+01,  2.37586031e+01,  3.23650484e+01,
        2.42125794e+01,  1.61880600e+01,  1.87279403e+01,  3.10467105e+01,
        1.58209827e+01,  2.39169645e+01,  2.70319068e+01,  2.00346550e+01,
        1.93732446e+01,  3.44786844e+01,  2.31527662e+01,  3.53225705e+01,
        2.26105398e+01,  1.43837813e+01,  2.66898889e+01,  1.58980379e+01,
        3.56766500e+01,  3.17486222e+01,  2.24073651e+01,  2.79839745e+01,
        1.55038571e+01,  3.87740295e+01,  2.77790326e+00,  3.99005215e-02,
        3.04058173e+01,  1.00014711e+01,  1.88600247e+01,  2.01080332e+01,
        2.84773446e+01,  8.81440292e+00,  1.92562123e+01,  1.17036958e+01,
        2.65171844e+01,  3.24174746e+00,  1.65123971e+01,  2.49366431e+01,
        2.26242956e+01,  2.11199715e+01,  2.46397951e+01,  4.01565956e+01,
        3.49345881e+01,  2.18745406e+01,  1.14354196e+01,  2.07182010e+01,
        1.27531309e+01,  1.95981203e+01,  1.06047576e+01,  2.92340621e+01,
        2.18628778e+01,  

Let's compare our predicted values to our ground truths

In [13]:
comp_df = pd.DataFrame({'y': y_val, 'y_pred': linearModelPredictions})
comp_df.head(n=10)

y     y_pred
ID                  
44   24.7  22.623571
472  19.6  23.421224
109  19.8  23.758603
293  27.9  32.365048
85   23.9  24.212579
458  13.5  16.188060
435  11.7  18.727940
267  30.7  31.046711
317  17.8  15.820983
297  27.1  23.916965

In all honesty, we shouldn't be using our eyes to judge. We should use a metric like MAE or RMSE to compare one model to the next! Let's import another function from scikit-learn to give us a hand

In [14]:
import math
from sklearn.metrics import mean_squared_error

In [15]:
mse = mean_squared_error(y_val, linearModelPredictions)
rmse = math.sqrt(mse)
print('RMSE = {}'.format(rmse))

RMSE = 5.321052760278853


Recall that this RMSE does not mean much right now. However, we can make it a baseline. Let's try to improve it.

Right now, what we have is a linear model. Let's build a polynomial model. What that means is, we will **engineer** new features. We will keep it simple by taking each feature and crossing it with only itself, essentially creating a square of each feature.

`sklearn` makes it easy to create squares of features. To make this work for us, we will make use of **pipelines**. Pipelines make it possible to apply a list of transformations. The particular transformation we will make use of is `PolynomialFeatures`.

## Polynomial Features

In [16]:
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

When we cross our features, we quickly run into very large numbers. To avoid having to deal with such large numbers, we will scale our features down to a range between 0 and 1. We will make use of `StandardScaler` to accomplish this.

Pipelines are made up of steps. We will define our transformations as steps. The steps we define are to scale our data, then create polynomials of degree 2, and then pass all of that to our Linear regression model.

In [17]:
steps = [
    ('scalar', StandardScaler()),
    ('poly', PolynomialFeatures(degree=2)),
    ('model', LinearRegression())
]

After defining our steps, we can then create our pipeline.

In [19]:
pipeline = Pipeline(steps)

At this point, we are ready to treat `pipeline` the same way we treat models. We can train by calling `fit`

In [20]:
pipeline.fit(X_train, y_train)

Pipeline(memory=None,
     steps=[('scalar', StandardScaler(copy=True, with_mean=True, with_std=True)), ('poly', PolynomialFeatures(degree=2, include_bias=True, interaction_only=False)), ('model', LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False))])

In [21]:
polyModelPredictions = pipeline.predict(X_val)

We can check our RMSE to see whether our model is getting better!

In [22]:
mse = mean_squared_error(y_val, polyModelPredictions)
rmse = math.sqrt(mse)
print('RMSE = {}'.format(rmse))

RMSE = 331.8541944770887


Our RMSE has got significantly worse! Let's check our scores.

In [23]:
print('Training score: {}'.format(pipeline.score(X_train, y_train)))

Training score: 0.9999468301332847


In [25]:
print('Test score: {}'.format(pipeline.score(X_val, y_val)))

Test score: -1302.1396137052345


From our training and test score above, we can tell that our model has overfit to the data. We need to introduce one or more tools to fix this.

## Fix Overfitting/Variance
The popular approaches to reducing variance are:
1. Add more data
1. Use data augmentation
1. Use architectures that generalize well
1. Add regularization
1. Reduce architecture complexity

The first two options are not available to us, so we will try using other architectures, adding regularization, and reducing architecture complexity. Scikit-learn provides three models that provide regularization:
* Lasso
* Ridge
* ElasticNet

These models also reduce complexity. `Lasso` reduces model complexity by completely eliminating certain features. `Ridge` reduces model complexity by driving the feature weights to a value very close to 0.

# Lasso

In [26]:
from sklearn.linear_model import Lasso, Ridge, ElasticNet

In [27]:
steps = [
    ('scalar', StandardScaler()),
    ('poly', PolynomialFeatures(degree=2)),
    ('model', Lasso(alpha=0.9))
]

lasso_pipe = Pipeline(steps)
lasso_pipe.fit(X_train, y_train)

Pipeline(memory=None,
     steps=[('scalar', StandardScaler(copy=True, with_mean=True, with_std=True)), ('poly', PolynomialFeatures(degree=2, include_bias=True, interaction_only=False)), ('model', Lasso(alpha=0.9, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False))])

In [28]:
lassoModelPredictions = lasso_pipe.predict(X_val)
mse = mean_squared_error(y_val, lassoModelPredictions)
rmse = math.sqrt(mse)
print('RMSE = {}'.format(rmse))

RMSE = 4.471162348530323


Would you take a look at that! Our RMSE has fixed itself and gone down. Our Lasso model has got rid of our overfitting. Lets take a look at our scores.

In [29]:
print('Training score: {}'.format(lasso_pipe.score(X_train, y_train)))

Training score: 0.8172033635115156


In [30]:
print('Test score: {}'.format(lasso_pipe.score(X_val, y_val)))

Test score: 0.763442237743289


Our training and test scores are now closer. Our overfitting problem has reduced. We can also improve our model by playing around with our regularization parameter `alpha`.

# Ridge

In [56]:
steps = [
    ('scalar', StandardScaler()),
    ('poly', PolynomialFeatures(degree=2)),
    ('model', Ridge(alpha=0.9))
]

ridge_pipe = Pipeline(steps)
ridge_pipe.fit(X_train, y_train)

Pipeline(memory=None,
     steps=[('scalar', StandardScaler(copy=True, with_mean=True, with_std=True)), ('poly', PolynomialFeatures(degree=2, include_bias=True, interaction_only=False)), ('model', Ridge(alpha=0.9, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001))])

In [57]:
ridgeModelPredictions = ridge_pipe.predict(X_val)
mse = mean_squared_error(y_val, ridgeModelPredictions)
rmse = math.sqrt(mse)
print('RMSE = {}'.format(rmse))

RMSE = 5.8229374742527495


For this particular dataset, `Ridge` is not as good as `Lasso`, but for other datasets, the situation might be different.

We will leave the implementation of `ElasticNet` to the student.

# TensorFlow

TensorFlow supports the creation of Linear Models using the `Estimator` API. Using TensorFlow requires some additional preparation because of the robust nature of solutions created using it. TensorFlow supports data that does not fit into your computer memory, and also supports training models over a distributed number of processors. At first, TensorFlow appears to be extremely verbose when compared to other libraries, but hang in there.

To keep things simple and interesting, we will make use of the same data above, which fits into memory. Linear models are created using `LinearRegressor`. We will need to do the following:
1. Create Feature Columns
1. Create Training, Validation, and Test Input Functions
1. Create Estimator
1. Train Estimator
1. Evaluate Estimator
1. Use Estimator

In [58]:
# import tensorflow
import tensorflow as tf

### Feature Columns

In [59]:
crim = tf.feature_column.numeric_column('crim', dtype=tf.float64, shape=())
zn = tf.feature_column.numeric_column('zn', dtype=tf.float64, shape=())
indus = tf.feature_column.numeric_column('indus', dtype=tf.float64, shape=())
chas = tf.feature_column.numeric_column('chas', dtype=tf.int64, shape=())
nox = tf.feature_column.numeric_column('nox', dtype=tf.float64, shape=())
rm = tf.feature_column.numeric_column('rm', dtype=tf.float64, shape=())
age = tf.feature_column.numeric_column('age', dtype=tf.float64, shape=())
dis = tf.feature_column.numeric_column('dis', dtype=tf.float64, shape=())
rad = tf.feature_column.numeric_column('rad', dtype=tf.int64, shape=())
tax = tf.feature_column.numeric_column('tax', dtype=tf.int64, shape=())
ptratio = tf.feature_column.numeric_column('ptratio', dtype=tf.float64, shape=())
black = tf.feature_column.numeric_column('black', dtype=tf.float64, shape=())
lstat = tf.feature_column.numeric_column('lstat', dtype=tf.float64, shape=())

feature_cols = [crim, zn, indus, chas, nox, rm, age, dis, rad, tax, ptratio, black, lstat]

### Input Functions

In [61]:
feature_names = ['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax', 'ptratio', 'black', 'lstat']
label_name = 'medv'

features_ndarray = boston_df[feature_names]
label_ndarray = boston_df[label_name]

X_train, X_test, y_train, y_test = train_test_split(features_ndarray, label_ndarray, random_state=0, test_size=0.3)

In [127]:
# training input function
def train_input():
    _dataset = tf.data.Dataset.from_tensor_slices(({'crim': X_train['crim'], 
                                                   'zn': X_train['zn'], 
                                                   'indus': X_train['indus'],
                                                   'chas': X_train['chas'],
                                                   'nox': X_train['nox'],
                                                   'rm': X_train['rm'],
                                                   'age': X_train['age'],
                                                   'dis': X_train['dis'],
                                                   'rad': X_train['rad'],
                                                   'tax': X_train['tax'],
                                                   'ptratio': X_train['ptratio'],
                                                   'black': X_train['black'],
                                                   'lstat': X_train['lstat']
                                                  }, y_train))
    dataset = _dataset.batch(16)
    iterator = dataset.make_one_shot_iterator()
    features, labels = iterator.get_next()
    return features, labels

In [128]:
# validation input function
def val_input():
    _dataset = tf.data.Dataset.from_tensor_slices(({'crim': X_test['crim'], 
                                                   'zn': X_test['zn'], 
                                                   'indus': X_test['indus'],
                                                   'chas': X_test['chas'],
                                                   'nox': X_test['nox'],
                                                   'rm': X_test['rm'],
                                                   'age': X_test['age'],
                                                   'dis': X_test['dis'],
                                                   'rad': X_test['rad'],
                                                   'tax': X_test['tax'],
                                                   'ptratio': X_test['ptratio'],
                                                   'black': X_test['black'],
                                                   'lstat': X_test['lstat']
                                                  }, y_test))
    dataset = _dataset.batch(16)
    iterator = dataset.make_one_shot_iterator()
    features, labels = iterator.get_next()
    return features, labels

### Create Estimator

When we create our `Estimator`, we have the option of specifying our optimizer. The optimizer determines how **gradient descent** is carried out during training. We are able to specify things like our `learning rate` using the optimizer. If we do not specify an optimizer, a default is used, along with a default learning rate.

In [137]:
optimizer = tf.train.FtrlOptimizer(learning_rate=0.1, l1_regularization_strength=0.9)
estimator = tf.estimator.LinearRegressor(feature_columns=feature_cols, optimizer=optimizer)

INFO:tensorflow:Using default config.
INFO:tensorflow:Using config: {'_model_dir': '/var/folders/2d/49w9fshj6ljb1kjldptz0w7h0000gn/T/tmpip0k2qmk', '_tf_random_seed': None, '_save_summary_steps': 100, '_save_checkpoints_steps': None, '_save_checkpoints_secs': 600, '_session_config': None, '_keep_checkpoint_max': 5, '_keep_checkpoint_every_n_hours': 10000, '_log_step_count_steps': 100, '_train_distribute': None, '_service': None, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x1a3790aa90>, '_task_type': 'worker', '_task_id': 0, '_global_id_in_cluster': 0, '_master': '', '_evaluation_master': '', '_is_chief': True, '_num_ps_replicas': 0, '_num_worker_replicas': 1}


### Train model

In [138]:
estimator.train(input_fn=train_input, steps=5000)

INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Saving checkpoints for 1 into /var/folders/2d/49w9fshj6ljb1kjldptz0w7h0000gn/T/tmpip0k2qmk/model.ckpt.
INFO:tensorflow:loss = 10394.72, step = 1
INFO:tensorflow:Saving checkpoints for 15 into /var/folders/2d/49w9fshj6ljb1kjldptz0w7h0000gn/T/tmpip0k2qmk/model.ckpt.
INFO:tensorflow:Loss for final step: 284.82892.


### Evaluate model

In [131]:
train_e = estimator.evaluate(input_fn=train_input)
test_e = estimator.evaluate(input_fn=val_input)

INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Starting evaluation at 2018-10-03-18:17:21
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from /var/folders/2d/49w9fshj6ljb1kjldptz0w7h0000gn/T/tmp01npc0w_/model.ckpt-15
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Finished evaluation at 2018-10-03-18:17:22
INFO:tensorflow:Saving dict for global step 15: average_loss = 99.76284, global_step = 15, loss = 1549.6495
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Starting evaluation at 2018-10-03-18:17:22
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from /var/folders/2d/49w9fshj6ljb1kjldptz0w7h0000gn/T/tmp01npc0w_/model.ckpt-15
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Finished evaluation at 2018-10-03-18:17:23
INFO:tensorflow:Saving dict for glob

In [132]:
print(train_e)

{'average_loss': 99.76284, 'loss': 1549.6495, 'global_step': 15}


In [133]:
print(test_e)

{'average_loss': 72.69981, 'loss': 1038.5687, 'global_step': 15}


### Predict

In [139]:
preds = estimator.predict(input_fn=val_input)
predictions = np.array([item['predictions'][0] for item in preds])
print(predictions)

INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from /var/folders/2d/49w9fshj6ljb1kjldptz0w7h0000gn/T/tmpip0k2qmk/model.ckpt-15
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
[20.092049  20.231483   6.9307528 16.890944  18.189425  17.180819
 17.975758   7.4399843 18.447449  18.416367  17.567406  18.895447
 22.23371   18.209578  18.928755  21.986542  18.023266  18.199894
 19.886023  19.127634  18.94459   18.019047  13.781283  18.381908
  1.8663982 18.048704  20.672987  24.861275  17.926315  18.481607
 18.71885   19.898882  16.923288  18.79068   18.12455   19.885622
 19.870726  16.906908  19.900043  18.466372  19.114716  22.94035
 20.678795  18.816011  19.837313  20.049385  19.070826  23.526049
 18.463333  13.484063  19.955206  17.966341  18.143766  14.274814
 18.398468  17.748072  18.452711  13.898594  25.591286  19.351114
 18.602228  24.615936  14.76788

In [140]:
mse = mean_squared_error(y_test, predictions)
rmse = math.sqrt(mse)
print('RMSE = {}'.format(rmse))

RMSE = 8.52641484618118
